analyzer/consensus/messages.go

```diff
@@ -0,0 +1,153 @@
+package consensus
+
+import (
+	"encoding/json"
+
+	apiTypes "github.com/oasisprotocol/nexus/api/v1/types"
+	"github.com/oasisprotocol/nexus/coreapi/v22.2.11/roothash/api/message"
+	"github.com/oasisprotocol/nexus/log"
+)
+
+type parsedMessage struct {
+	messageType      string
+	body             json.RawMessage
+	relatedAddresses map[apiTypes.Address]struct{}
+}
+
+func extractMessageData(logger *log.Logger, messageRaw json.RawMessage) parsedMessage {
+	pm := parsedMessage{
+		relatedAddresses: map[apiTypes.Address]struct{}{},
+	}
+	var m message.Message
+	if err := json.Unmarshal(messageRaw, &m); err != nil {
```

Comment:
oh yeah I think we can do that now that we have eden stuff backported into our damask/universal api

Reply:
or should we, this is kind of a parallel to transactions, and for transactions we leave them in a non-struct from in the nodeapi, as just a signed transaction. then the analyzer layer does the conversion and switching on method type

Reply:
hm if we follow this second pattern, wouldn't that mean keeping messages in their CBOR form when they come from nodeapi, and then converting them directly to message.Message? I'm just worried that the extra JSON (de)serialization might introduce subtle changes to the data (e.g. around large numbers), so I'd rather avoid it.

Your call on which way to go.

Reply:
a few options:
1. leave it as []json.RawMessage
  a. (currently does this) reserialize vendored message.Message into JSON. extra serialization round trip
2. change it to []cbor.RawMessage
  a. reserialize vendored message.Message into CBOR instead of into JSON, but it would still be a round trip
  b. patch vendored message.Message into cbor.RawMessage, avoids round trip, probably not worth it
3. change nodeapi to use damask message.Message
  a. write a convertRoothashMessage suite basically to copy things over field-for-field
  b. reserialize into JSON and deserialize into damask message.Message
  c. reserialize into CBOR and deserialize into damask message.Message
  d. patch vendored message.Message into cbor.RawMessage and deserialize into damask message.Message
  e. patch vendored message.Message into damask message.Message

Reply:
3c is striking me as pretty un-dystopian

Reply:
huh, unrelated (?) diff in emerald

```
--- a/tmp/nexus-e2e-expected-eden/db__contract_gas_use.csv
+++ b/home/wh0/work/ekiden/explorer/indexer/tests/e2e_regression/eden/actual/db__contract_gas_use.csv
@@ -93,6 +93,7 @@ emerald,oasis1qzr3tn5zg0cu4kzd7jar0hx8zeg77xqjavrqst0k,471113,
 emerald,oasis1qzv3xvhvf67g5tpck5fy2mr3pndavnaptypm8vrv,,
 emerald,oasis1qzvvs98fj36nftgu3tdwwq3wrcx7y99t0cpngwrd,,
 emerald,oasis1qzwr3xtmuq0q0rufe54xcl7ajrlghu7fgqlkvvue,146586,
+emerald,oasis1qzwytggxna94zgqpq7wutcc79f9gdyj5mcu49wtg,,
 emerald,oasis1qzxcveap204cg0g962nyrh864nuxpzgmrs2sv9r4,,
 emerald,oasis1qzxe8q6s3xmqt26dlwf905e8t9szfydgmqa062en,,
 emerald,oasis1qzy65m52h5pz5z7vmr08n3jr23v7cjn67vvfwjys,781999,
diff --git a/tmp/nexus-e2e-expected-eden/db__evm_contracts.csv b/home/wh0/work/ekiden/explorer/indexer/tests/e2e_regression/eden/actual/db__evm_contracts.csv
index a5231391..b34ec1d7 100644
--- a/tmp/nexus-e2e-expected-eden/db__evm_contracts.csv
+++ b/home/wh0/work/ekiden/explorer/indexer/tests/e2e_regression/eden/actual/db__evm_contracts.csv
@@ -93,6 +93,7 @@ emerald,oasis1qzr3tn5zg0cu4kzd7jar0hx8zeg77xqjavrqst0k,,
 emerald,oasis1qzv3xvhvf67g5tpck5fy2mr3pndavnaptypm8vrv,,83a097b551dd28242d25a9836d3c18de
 emerald,oasis1qzvvs98fj36nftgu3tdwwq3wrcx7y99t0cpngwrd,,31f08a1ebc85d12804ece728924c376e
 emerald,oasis1qzwr3xtmuq0q0rufe54xcl7ajrlghu7fgqlkvvue,,
+emerald,oasis1qzwytggxna94zgqpq7wutcc79f9gdyj5mcu49wtg,,bd61c9b8da1b2ad026662c2e5adbcac1
 emerald,oasis1qzxcveap204cg0g962nyrh864nuxpzgmrs2sv9r4,,37f73a0d935e20288628373578da84c2
 emerald,oasis1qzxe8q6s3xmqt26dlwf905e8t9szfydgmqa062en,,6fcd54561cd3a5d5db380b21462ac0b6
 emerald,oasis1qzy65m52h5pz5z7vmr08n3jr23v7cjn67vvfwjys,,
```

Reply:
e2e diff: hrmph my guess is that this contract became newly verified on sourcify, and we pulled it in unrelated to your changes. Maybe because you invalidated your caches too aggresively? You could try restoring them and running e2e again. Or maybe we have a bug and our sourcify client sometimes bypasses cache and hits up the real sourcify even when a cached response is available?

Reply:
> 3c is striking me as pretty un-dystopian
 
Is this an endorsement? Yeah I can get behind 3c :). I'd probably go for 3a myself because it's what we do for events and txs, and having done it there, it's less work than it seems originally. Lots of lines of code, but quick stupid boilerplate, especially with good autocomplete. And it gives us good (in)compatibility warnings between versions. OTOH it does feel like slightly wasted effort given the messages' lower prominence/importance.

Reply:
ya, probably don't wanna go full utopian, just non-dystopian

Reply:
implemented 3c